google-cloud-spanner/src/test/java/com/google/cloud/spanner/it/ITDirectPathFallback.java

```diff
@@ -0,0 +1,305 @@
+
+  @ClassRule public static IntegrationTestEnv env = new IntegrationTestEnv();
+
+  private AtomicBoolean blackholeDpAddr = new AtomicBoolean();
+  private AtomicInteger numBlocked = new AtomicInteger();
+  private AtomicInteger numDpAddrRead = new AtomicInteger();
+  private boolean isDpAddr;
+
+  private ChannelFactory<NioSocketChannel> channelFactory;
+  private EventLoopGroup eventLoopGroup;
+  private RemoteSpannerHelper testHelper;
+
+  private static final String TABLE_NAME = "TestTable";
+  private static final List<String> ALL_COLUMNS = Arrays.asList("Key", "StringValue");
+  private static Database db;
+  private static DatabaseClient client;
+
+  private static final String DIRECT_PATH_ENDPOINT = "aa423245250f2bbf.sandbox.googleapis.com:443";
+  private static final String ATTEMPT_DIRECT_PATH = "spanner.attempt_directpath";
+
+  public ITDirectPathFallback() {
+    // Create a transport channel provider that can intercept ipv6 packets.
+    channelFactory = new MyChannelFactory();
+    eventLoopGroup = new NioEventLoopGroup();
+  }
+
+  @Before
+  public void setup() throws IOException, Throwable {
+    assume()
+        .withMessage("DirectPath integration tests can only run against DirectPathEnv")
+        .that(Boolean.getBoolean(ATTEMPT_DIRECT_PATH))
+        .isTrue();
+    // Get default spanner options for Ingetration test
+    SpannerOptions.Builder builder = env.getTestHelper().getOptions().toBuilder();
+    // Set instrumented transport provider
+    builder.setChannelProvider(
+        InstantiatingGrpcChannelProvider.newBuilder()
+            .setAttemptDirectPath(true)
+            .setEndpoint(DIRECT_PATH_ENDPOINT)
+            .setPoolSize(1)
+            .setChannelConfigurator(
+                new ApiFunction<ManagedChannelBuilder, ManagedChannelBuilder>() {
+                  @Override
+                  public ManagedChannelBuilder apply(ManagedChannelBuilder builder) {
+                    injectNettyChannelHandler(builder);
+                    // Fail fast when blackhole is active
+                    builder.keepAliveTime(1, TimeUnit.SECONDS);
+                    builder.keepAliveTimeout(1, TimeUnit.SECONDS);
+                    return builder;
+                  }
+                })
+            .build());
+    // Forcefully ignore GOOGLE_APPLICATION_CREDENTIALS
+    builder.setCredentials(
+        FixedCredentialsProvider.create(ComputeEngineCredentials.create()).getCredentials());
+
+    // Create a new testHelper with the instrumented transport provider
+    testHelper = RemoteSpannerHelper.create(builder.build(), env.getTestHelper().getInstanceId());
+
+    db =
+        testHelper.createTestDatabase(
+            "CREATE TABLE TestTable ("
+                + "  Key                STRING(MAX) NOT NULL,"
+                + "  StringValue        STRING(MAX),"
+                + ") PRIMARY KEY (Key)",
+            "CREATE INDEX TestTableByValue ON TestTable(StringValue)",
```

Comment:
nit: Could we remove these indexes? I don't think they are relevant for this test, and creating and maintaining secondary indexes are expensive operations.

Reply:
Done. Thanks!